nixos/modules/services/security/certmgr.nix

```diff
@@ -0,0 +1,210 @@
+        else [""])
+      ) set));
+
+  # If cfg.mkTargetDirs is true,
+  # .. select all attributes named "path" under declSpecs and resolve their directories for mkdir
+  preStart = concatStringsSep "\n" [
+    (optionalString cfg.mkTargetDirs
+      (concatStringsSep " "
+        (["mkdir -p"] ++ (uniqueFilter (catDirsRecursive (n: v: n == "path") cfg.declSpecs)))
+      )
+    )
+    ("${pkgs.certmgr}/bin/certmgr -f ${certmgrYaml} check")
+  ];
+in
+{
+  options.services.certmgr = {
+    enable = mkEnableOption "certmgr";
+
+    defaultRemote = mkOption {
+      type = types.str;
+      default = "127.0.0.1:8888";
+      description = "The default CA host:port to use.";
+    };
+
+    before = mkOption {
+      default = "72h";
+      type = types.str;
+      description = "The interval before a certificate expires to start attempting to renew it.";
+    };
+
+    interval = mkOption {
+      default = "30m";
+      type = types.str;
+      description = "How often to check certificate expirations and how often to update the cert_next_expires metric.";
+    };
+
+    metricsAddress = mkOption {
+      default = "127.0.0.1";
+      type = types.str;
+      description = "The address for the Prometheus HTTP endpoint.";
+    };
+
+    metricsPort = mkOption {
+      default = 9488;
+      type = types.ints.u16;
+      description = "The port for the Prometheus HTTP endpoint.";
+    };
+
+    impSpecs = mkOption {
+      default = [];
+      example = singleton {
+        name = "example-cert";
+        path = "/var/lib/certmgr/specs/example-cert.json";
+      };
+      type = with types; listOf (submodule {
+        options.name = mkOption { type = str; description = "name of the symlink.";   };
+        options.path = mkOption { type = path; description = "target of the symlink."; };
+      });
+      description = "List of { name, path } to link into the specsdir imperatively.";
+    };
+
+    declSpecs = mkOption {
+      default = {};
+      example = literalExample ''
+      {
+        exampleCert =
+        let
+          domain = "example.com";
+          secret = name: "/var/lib/secrets/''${name}.pem";
+        in {
+          service = "nginx";
+          action = "reload";
+          authority = {
+            file.path = secret "ca";
+          };
+          certificate = {
+            path = secret domain;
+          };
+          private_key = {
+            owner = "root";
+            group = "root";
+            mode = "0600";
+            path = secret "''${domain}-key";
+          };
+          request = {
+            CN = domain;
+            hosts = [ "mail.''${domain}" "www.''${domain}" ];
+            key = {
+              algo = "rsa";
+              size = 2048;
+            };
+            names = {
+              O = "Example Organization";
+              C = "USA";
+            };
+          };
+        };
+      }
+      '';
+      type = with types; (attrsOf (submodule ({ ... }: {
```

Comment:
Removed parens as part of: https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b .. Will look into the possibility (and potential problems) of merging impSpecs and declSpecs now.

Reply:
Combined `impSpecs` and `declSpecs` in https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2

Reply:
While I understand the intention behind combining the specs, I think it'll lead to a poorer understanding of the intended separation. It's important that `impSpec`s are never imported into the store, since they'll most likely (always?) contain secrets. The `either` machinery necessary to facilitate a joined `specs` introduces a lot of opportunities for future maintainers to accidentally import the paths -- and even if it's safe (which it is in the current state) the error messages if you muck up the value are not great.

In contrast, `impSpecs` might look non-dry, but it requires very trivial machinery and it'll be easier to ensure that none of its values are accidentally leaked into the store.